[LLD] Implement --enable-non-contiguous-regions

mysterymath opened the pull request:
When enabled, input sections that would otherwise overflow a memory region are instead spilled to the next matching output section.

This feature parallels the one in GNU LD, but there are some differences from its documented behavior:

- /DISCARD/ only matches previously-unmatched sections (i.e., the flag does not affect it).

- If a section fails to fit at any of its matches, the link fails instead of discarding the section.

- The flag --enable-non-contiguous-regions-warnings is not implemented, as it exists to warn about such occurrences.

The implementation places stubs at possible spill locations, and replaces them with the original input section when effecting spills. Spilling decisions occur after address assignment. Sections are spilled in reverse order of assignment, with each spill naively decreasing the size of the affected memory regions. This continues until the memory regions are brought back under size. Spilling anything causes another pass of address assignment, and this continues to fixed point.

Spilling after rather than during assignment allows the algorithm to consider the size effects of unspillable input sections that appear later in the assignment. Otherwise, such sections (e.g. thunks) may force an overflow, even if spilling something earlier could have avoided it.

A few notable feature interactions occur:

- Stubs affect alignment, ONLY_IF_RO, etc, broadly as if a copy of the input section were actually placed there.

- SHF_MERGE synthetic sections use the spill list of their first contained input section (the one that gives the section its name).

- ICF occurs oblivious to spill sections; spill lists for merged-away sections become inert and are removed after assignment.

- SHF_LINK_ORDER and .ARM.exidx are ordered according to the final section ordering, after all spilling has completed.

- INSERT BEFORE/AFTER and OVERWRITE_SECTIONS are explicitly disallowed.

mysterymath committed:
[LLD] Implement --enable-non-contiguous-regions

When enabled, input sections that would otherwise overflow a memory
region are instead spilled to the next matching output section.

This feature parallels the one in GNU LD, but there are some differences
from its documented behavior:

- /DISCARD/ only matches previously-unmatched sections (i.e., the flag
  does not affect it).

- If a section fails to fit at any of its matches, the link fails
  instead of discarding the section.

- The flag --enable-non-contiguous-regions-warnings is not implemented,
  as it exists to warn about such occurrences.

The implementation places stubs at possible spill locations, and
replaces them with the original input section when effecting spills.
Spilling decisions occur after address assignment. Sections are spilled
in reverse order of assignment, with each spill naively decreasing the
size of the affected memory regions. This continues until the memory
regions are brought back under size. Spilling anything causes another
pass of address assignment, and this continues to fixed point.

Spilling after rather than during assignment allows the algorithm to
consider the size effects of unspillable input sections that appear
later in the assignment. Otherwise, such sections (e.g. thunks) may
force an overflow, even if spilling something earlier could have avoided
it.

A few notable feature interactions occur:

- Stubs affect alignment, ONLY_IF_RO, etc, broadly as if a copy of the
  input section were actually placed there.

- SHF_MERGE synthetic sections use the spill list of their first
  contained input section (the one that gives the section its name).

- ICF occurs oblivious to spill sections; spill lists for merged-away
  sections become inert and are removed after assignment.

- SHF_LINK_ORDER and .ARM.exidx are ordered according to the final
  section ordering, after all spilling has completed.

- INSERT BEFORE/AFTER and OVERWRITE_SECTIONS are explicitly disallowed.

smithp35 commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1364,6 +1432,107 @@ const Defined *LinkerScript::assignAddresses() {
+        spilled = true;
+        SpillList &list = it->second;
+
+        SpillInputSection *spill = list.head;
```
Would it be worth using `potentialSpill` here. We have `spill` `spills`, `spillLists` all with fairly close names. I found it hard to keep track of what was what on first reading.

mysterymath replied:
I think `spills` is the misnamed one here; `spill` is a potential spill that is currently being taken, but `spills` isn't a list of such things; it's a list of spilled input sections. Added a comment for the variable and renamed it to `spilledInputSections`.

smithp35 commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1364,6 +1432,107 @@ const Defined *LinkerScript::assignAddresses() {
+          continue;
+
+        spilled = true;
+        SpillList &list = it->second;
```
list is very generic. Would it be worth something like
`isecSpillLocations` or `isecSpillLocList`?

mysterymath replied:
Hmm, this is just a stylistic preference on my end, but I tend to use small anaphora (it for iterator, list for a list, etc.) for tightly-scoped variables that temporarily give names to objects before disappearing. Here there's only one list in question (the spill list for the input section), and the scope spans only five lines.

As with all things stylistic, there's tons of nuance, and if you did find this difficult to read, knowing more about why would help me refine the rule.

smithp35 replied:
It was a similar situation to a previous comment, there were several other lists, spills and spilllist and I found myself having to double check which was which.

I don't have a particularly strong opinion though, so feel free to ignore me here.

smithp35 commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1364,6 +1432,107 @@ const Defined *LinkerScript::assignAddresses() {
+        // its properties to match the new location.
+        *llvm::find(spill->isd->sections, spill) = isec;
+        isec->parent = spill->parent;
+        // The alignment of the spill section may have diverged from the
```
Presumably this might be increased due to some linker script expression like subalign?

mysterymath replied:
Yep, that's exactly the case that came up when hammering on the tests. Probably worth explicitly calling out in a comment.

smithp35 commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1364,6 +1432,107 @@ const Defined *LinkerScript::assignAddresses() {
+        spills.insert(isec);
+
+        // Replace the next spill location with the spilled section and adjust
+        // its properties to match the new location.
```
When spilling we are moving OutputSections and this might affect the flags of the output section.

Do we account for splill placements that might have awkward combinations of flags?

I can't at a glance see commitSection() being called https://github.com/llvm/llvm-project/blob/main/lld/ELF/OutputSections.cpp#L116 but I could be missing something.

mysterymath replied:
This was one of the semantic caveats taken to keep the model of how this works simple. The spill input sections work throughout the whole link as if there were additional copies of the referenced input section in those output sections. This includes calculations for the flags of the output sections, program header generation, etc. The one caveat is they're not included in the list of all input sections, so they're not in scope for ICF, GC, etc.
 
It's not until the final address assignment in fixed point that the copies drop out. Afterwards, the input sections are seen in their final positions, and the spills have been removed, albeit leaving some of their effects behind.

This isn't as simple as possible; the simplest semantics would be for the whole link to behave "as if" the sections had always been matched to their final positions. However, due to circular dependencies in the link, this would require either running huge swaths of the linker in a loop to fixed point or being able to undo and redo chains of decision-making between output section matching and assignment. The former seems broadly like a non-starter, but the latter we may be able to do piecemeal if the semantics above don't actually end up being sufficient.

smithp35 replied:
OK I see, thanks.

smithp35 commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1364,6 +1432,107 @@ const Defined *LinkerScript::assignAddresses() {
+        isec->addralign = spill->addralign;
+
+        // Record the reduction in overage.
+        osec->memRegion->curPos -= isec->getSize();
```
In theory this could be a potentially large underestimate if isec has a high alignment and there is a lot of alignment padding.

mysterymath replied:
That's true; this is definitely a best-effort satisficing type of algorithm.

mysterymath replied:
That's true; this is definitely a best-effort satisficing type of algorithm.

The situation could be improved by running the spilling both forwards during assignment and backwards afterwards (as done here). This could be improved further binary searching for the earliest "split point" in the address assignment that allows the first memory region to fit, then fixing that moving on to the next overflowing region, etc, for `O(m log n)` time, where `m = |regions|` and `n = |sections|`. Alternatively, we may be able to handle alignment more specifically and accurately determine its effects. At the extreme end, we could get a kind of optimality by backtracking whenever a forward assignment overflows and performing the last spill opportunity, for `O(n^2)` time. That might not actually be too bad in practice.

Anyway, I do think that there's potential room to grow here, but my goal was to find the simplest algorithm that might satisfice most embedded projects. I had originally only had forward spilling (inline with assignment), but late unspillable sections cropped up too many times, and backwards spilling is resistant to them (while having its own concerns).

smithp35 replied:
Agreed it is best to keep it simple at first. I'm hoping that the majority of cases where a really highly aligned section is used will be specific to a single region.

smithp35 reviewed (commented):
Thanks very much for working on this. I've not had a chance to work through the tests yet.

So far most comments are cosmetic so I think this is looking good from my size. I expect MaskRay will want to look at the code in detail.

mysterymath committed:
Address round of review comments

- Add no-op `--enable-non-contiguous-regions-warnings`.
- Add help text to LLD man page.
- Additional implementation comments.
- Remove dead variable.
- Remove release note merge chaff.
- Rename `isRegionOverflowed` to `hasRegionOverflowed`.
- Rename variable `is` to `isd` for InputSectionDescription.

mysterymath commented on lld/ELF/Options.td:
```diff
@@ -196,6 +196,9 @@ def emit_relocs: F<"emit-relocs">, HelpText<"Generate relocations in output">;
 def enable_new_dtags: F<"enable-new-dtags">,
   HelpText<"Enable new dynamic tags (default)">;
 
+def enable_non_contiguous_regions : FF<"enable-non-contiguous-regions">,
```
Ah, good idea; removes a needless compatibility ding. Done.

MaskRay replied:
We used to add some compatibility options or aliases, but mostly stopped doing do.
`--enable-non-contiguous-regions` is itself quite niche. `--enable-non-contiguous-regions-warnings` seems more niche and probably doesn't justify an ignored option.

mysterymath commented on lld/ELF/Options.td:
```diff
@@ -196,6 +196,9 @@ def emit_relocs: F<"emit-relocs">, HelpText<"Generate relocations in output">;
 
+def enable_non_contiguous_regions : FF<"enable-non-contiguous-regions">,
+  HelpText<"Spill input sections to later matching output sections to avoid memory region overflow">;
+
```
Done.

mysterymath commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -555,6 +583,28 @@ LinkerScript::computeInputSections(const InputSectionDescription *cmd,
   // input order).
   sortByPositionThenCommandLine(sizeAfterPrevSort, ret.size());
+
+  // Replace matches after the first with potential spill sections.
```
Tweaked the wording to be a bit more imperative; done.

mysterymath commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1066,8 +1126,16 @@ void LinkerScript::assignOffsets(OutputSection *sec) {
     // updates the output section size.
-    for (InputSection *isec : cast<InputSectionDescription>(cmd)->sections) {
+
+    DenseSet<InputSection *> spills;
```
Thanks, left over from a previous algorithm.

mysterymath commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1066,8 +1126,16 @@ void LinkerScript::assignOffsets(OutputSection *sec) {
+    for (InputSection *isec : sections) {
       assert(isec->getParent() == sec);
+
+      // Skip all possible spills.
```
Done.

mysterymath commented on lld/ELF/OutputSections.cpp:
```diff
@@ -226,6 +226,8 @@ void OutputSection::finalizeInputSections() {
         i = std::prev(mergeSections.end());
         syn->entsize = ms->entsize;
         isd->sections.push_back(syn);
+        if (script)
```
Done.

MaskRay replied:
`script` is always non-null, so you can pass a reference

MaskRay replied:
Since copyPotentialSpillList is so simple and used once, the function should be removed.

mysterymath commented on lld/docs/ReleaseNotes.rst:
```diff
@@ -35,6 +35,17 @@ ELF Improvements
+  relocatable object file. (`D146778 <https://reviews.llvm.org/D146778>`_)
+* common-page-size can now be larger than the system page-size.
+  (`#57618 <https://github.com/llvm/llvm-project/issues/57618>`_)
+* ``--enable-non-contiguous-regions`` option allows automatically packing input
```
Nope, must have come in from a merge somehow. Thanks.

mysterymath commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1364,6 +1432,107 @@ const Defined *LinkerScript::assignAddresses() {
   return getChangedSymbolAssignment(oldValues);
 }
 
+static bool isRegionOverflowed(MemoryRegion *mr) {
```
Done.

mysterymath commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1364,6 +1432,107 @@ const Defined *LinkerScript::assignAddresses() {
+          !isRegionOverflowed(osec->lmaRegion))
+        break;
+
+      auto *is = dyn_cast<InputSectionDescription>(cmd);
```
Done.

mysterymath commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1364,6 +1432,107 @@ const Defined *LinkerScript::assignAddresses() {
+        // not the original.
+        isec->addralign = spill->addralign;
+
+        // Record the reduction in overage.
```
Hah, this might be one of those American-English-isms. Reworked the wording and filled out this comment a bit more to make it clear why this is happening.

mysterymath commented on lld/ELF/InputSection.h:
```diff
@@ -425,6 +426,25 @@ class InputSection : public InputSectionBase {
+// A marker for a potential spill location for another input section. This
+// broadly acts as if it were the original section until address assignment.
+// Then it is either replaced with the real input section or removed.
+class SpillInputSection : public InputSection {
```
I think I like `PotentialSpillSection` better than `SpillInputSection`; it's not really an `InputSection`, and it's more like a `SyntheticSection`. That also gives `potentialSpillLists` for the the tracking variable and `PotentialSpillList` for its type, which is better than `spillLists` and `SpillList`.

mysterymath committed:
Rename SpillInputSection (etc.) to PotentialSpillSection

smithp35 commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions-arm-exidx.test:
```diff
@@ -0,0 +1,54 @@
+# If a spilling reorders input sections, the .ARM.exidx table must be rebuilt
```
There's also an unwritten convention in LLD to use an extra comment character for non FileCheck commands. For example:
```## If a spilling reorders ...```

It has been some time since that was started so could be worth looking at recent commits to see if it is still in operation.

smithp35 commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions.test:
```diff
@@ -0,0 +1,279 @@
+SPILL:      .first_chance PROGBITS 0000000000000000 001000 000001
+SPILL-NEXT: .last_chance  PROGBITS 0000000000000002 001002 000002
+
+# A spill off the end must still fails the link.
```
Typo?
`must still fail the link.`

mysterymath replied:
Fixed

smithp35 reviewed (commented):
Thanks very much for the update. Only a few small comments on the tests.

I definitely would like to see this feature added as the equivalent feature in Arm's proprietary toolchain has been widely used. Given that the equivalent GNU ld feature is new, I don't think we'll see existing widespread use but I think it could pick up in the future.

mysterymath committed:
Address exidx test review comments

- Update example to demonstrate merging
- Make test require ARM
- Two hashes for comments; one for non-split-file content
- Fixed typo

mysterymath commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions-arm-exidx.test:
```diff
@@ -0,0 +1,54 @@
+# If a spilling reorders input sections, the .ARM.exidx table must be rebuilt
```
Yep, thanks.

mysterymath commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions-arm-exidx.test:
```diff
@@ -0,0 +1,54 @@
+ .fnstart
+ bx lr
+ .save {r9, lr}
+ .setfp r9, sp, #0
```
> I think it could be worth making the unwind instructions identical for at least two of the sections. The .exidx table compression merges identical entries. One of things I was worried about with this patch was whether the table compression would still work properly when the order of sections could change.

Ah, good point. I had just copied this from another test, since I haven't learned the details of `.ARM.exidx`'s encoding yet. I've made two of the sets of assembly declarations identical and updated the assertions. Found that llvm-readobj had a nice pretty-printer for the exidx section, and the results seem plausible.

> However to make a test that actually fails to converge would be difficult and not representative of real world usage.
Even if the `.ARM.exidx` alternatively grows or shrinks on different passes, I'd think it still wouldn't affect convergence. At least, it shouldn't on the basis of spilling alone, since performing a spill is still an irrevocable action.

smithp35 reviewed (approved):
Thanks for the updated tests. I've checked that with a slightly different .ARM.exidx (changing to r7 rather than r8), this isn't prematurely compressed in the first pass.

Given MaskRay's comment in https://discourse.llvm.org/t/rfc-lld-enable-non-contiguous-regions/76513/17 it will be worth running this past him first before merging.

MaskRay commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions.test:
```diff
@@ -0,0 +1,279 @@
+# REQUIRES: x86
+
+# RUN: split-file %s %t
```
`RUN: rm -rf %t && split-file %s %t && cd %t`

The convention is to remove %t to avoid prevent lingering files. `cd %t` allows to remove `%t/` below, making RUN lines cleaner.

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -505,12 +510,33 @@ LinkerScript::computeInputSections(const InputSectionDescription *cmd,
+      if (!sec->isLive() || seen.contains(i))
         continue;
 
+      if (sec->parent) {
```
If this is moved before `ret.push_back(sec);`, we can save one `if (sec->parent)` check and group enableNonContiguousRegions code together.

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -555,6 +583,32 @@ LinkerScript::computeInputSections(const InputSectionDescription *cmd,
+
+      PotentialSpillSection *pss = make<PotentialSpillSection>(
+          sec, const_cast<InputSectionDescription *>(cmd));
+
```
The code style is pretty dense. Variables are not followed by a blank line.

`auto [it, inserted] = ..` is preferred for newer code

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1364,6 +1436,117 @@ const Defined *LinkerScript::assignAddresses() {
+
+        spilled = true;
+        PotentialSpillList &list = it->second;
+
```
This blank line and the next one are unneeded.

The code style is pretty dense. We usually insert blank lines for a larger block of code.

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1364,6 +1436,117 @@ const Defined *LinkerScript::assignAddresses() {
+      }
+
+      // Remove any spilled input sections to complete their move.
+      if (!spilledInputSections.empty())
```
`spilled = true` can be moved here to remove one line from the above code that requires more attention.

MaskRay commented on lld/ELF/Writer.cpp:
```diff
@@ -1464,6 +1466,12 @@ template <class ELFT> void Writer<ELFT>::finalizeAddressDependentContent() {
+    } else if (spilled) {
+      // Spilling can change relative section order, so recompute anything that
+      // depends on it.
+      for (Partition &part : partitions)
```
Create a lambda to avoid duplicating the few lines. We might add more in the future and could accidentally break --enable-non-contiguous-regions

MaskRay commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions.test:
```diff
@@ -0,0 +1,279 @@
+
+## A spill off the end must still fail the link.
+
+# RUN: not ld.lld -T %t/spill-fail.ld %t/spill.o -o %t/spill-fail --enable-non-contiguous-regions 2>&1 |\
```
For `not ld.lld`, we use `-o /dev/null`. But if `rm -rf %t && split-file %s %t && cd %t` is used, `-o` can be omitted.

ditto elsewhere

mysterymath replied:
Done via the latter.

MaskRay commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions.test:
```diff
@@ -0,0 +1,279 @@
+## A spill off the end must still fail the link.
+
+# RUN: not ld.lld -T %t/spill-fail.ld %t/spill.o -o %t/spill-fail --enable-non-contiguous-regions 2>&1 |\
+# RUN:   FileCheck %s -check-prefix=SPILL-FAIL
```
`--implicit-check-not=error:`

We used to have issues that linker script diagnostics are repeated, so it's worth enhancing the test.

mysterymath replied:
Done.

MaskRay commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions.test:
```diff
@@ -0,0 +1,279 @@
+# RUN: split-file %s %t
+# RUN: llvm-mc -n -filetype=obj -triple=x86_64 %t/spill.s -o %t/spill.o
+
+## An input section must spill to a later match if the region of its first match
```
There are many "must"s in comments. Our convention states what happens and omits "must". Very occasionally we use "should"

mysterymath replied:
Done; this is terser than I was able to come up with, which I appreciate.

MaskRay commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions.test:
```diff
@@ -0,0 +1,279 @@
+
+# OVERWRITE_SECTIONS: error: OVERWRITE_SECTIONS cannot be used with --enable-non-contiguous-regions
+
+# SHF_LINK_ORDER must be reordered if spilling changes relative section order.
```
`## `

mysterymath replied:
Thanks. Hah, I missed this while scanning because it was all caps after the hash.

MaskRay commented on lld/docs/ReleaseNotes.rst:
```diff
@@ -35,6 +35,12 @@ ELF Improvements
+  region would overflow. This reduces the toil of manually packing regions
+  (typical for embedded). It also makes full LTO feasible in such cases, since
+  IR merging currently prevents the linker script from referring to input
+  files.
```
``(`#90007 <https://github.com/llvm/llvm-project/pull/90007>`_)``

mysterymath replied:
Done.

MaskRay commented on lld/ELF/InputSection.cpp:
```diff
@@ -309,6 +310,12 @@ std::string InputSectionBase::getObjMsg(uint64_t off) const {
       .str();
 }
 
+PotentialSpillSection::PotentialSpillSection(InputSectionBase *source,
```
We prefer `const &`  when a parameter is non-null.

mysterymath replied:
Done.

MaskRay commented on lld/ELF/InputSection.h:
```diff
@@ -48,7 +48,7 @@ template <class ELFT> struct RelsOrRelas {
 class SectionBase {
 public:
-  enum Kind { Regular, Synthetic, EHFrame, Merge, Output };
+  enum Kind { Regular, Synthetic, EHFrame, Merge, Output, Spill };
```
`Regular, Synthetic, Spill, EHFrame, Merge, Output` may be more efficient.

mysterymath replied:
Done.

mysterymath committed:
Address review comments

- Remove split-file directory and cd into it.
- Consolidate logic in computeInputSections().
- Make spill logic denser and group comments.
- Create finalizeOrderDependentContent lambda.
- Spill zero-size sections.
- Remove warnings flag.
- Make potentialSpillLists public; remove copySpillLists.
- implicit-check-not: error:
- Reword test descriptions to remove "must".
- Merge spill and subalign test.
- Missing hash before SHF_LINK_ORDER comment.
- Add PR reference to release notes
- Take potentialSpillSection constructor params by ref

mysterymath commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions.test:
```diff
@@ -0,0 +1,279 @@
+
+## A spill must use the alignment of the later match.
+
+# RUN: ld.lld -T %t/spill-align.ld %t/spill.o -o %t/spill-align --enable-non-contiguous-regions
```
Done.

mysterymath commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions.test:
```diff
@@ -0,0 +1,279 @@
+.byte 1
+
+.section .link_order.b,"ao",@progbits,.b
+.byte 2
```
I'm not sure exactly what you mean here; the order swaps from the one given in the input file (.link_order.a, .link_order.b) to respond to the final order of input sections due to spilling: (.b, .a). Is there a combination of link order sections and spilling sections that would be less ambiguous or clearer?

MaskRay replied:
The idea is to have 3 SHF_LINK_ORDER sections to make sorting requirement clearer.

With two sections it's a bit less clear whether sorting is intended (it could be a reversed order).

mysterymath replied:
Ah, gotcha. Done.

mysterymath committed:
Review comments

- Swap true and false branches to remove condition negation
- No -o for failure tests
- Add a third link order section to show both normal sorting and
  reordering due to spilling.

mysterymath committed:
Remove trivial comment and blank lines; remove dead function declaration

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -1364,6 +1423,114 @@ const Defined *LinkerScript::assignAddresses() {
+          continue;
+        PotentialSpillList &list = it->second;
+        PotentialSpillSection *spill = list.head;
+        if (!spill->next)
```
Prefer `if (xxx)` to `if (!xxx)` by swapping then/else branches.

MaskRay commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions.test:
```diff
@@ -0,0 +1,259 @@
+
+## An additional match in /DISCARD/ has no effect.
+
+# RUN: not ld.lld -T no-spill-into-discard.ld spill.o -o no-spill-into-discard --enable-non-contiguous-regions 2>&1 |\
```
`not ld.lld` commands can omit  `-o` (default: `a.out`) if `cd %t` has been invoked.

MaskRay commented on lld/test/ELF/linkerscript/enable-non-contiguous-regions.test:
```diff
@@ -0,0 +1,259 @@
+
+## An error is reported for OVERWRITE_SECTIONS.
+
+# RUN: not ld.lld -T overwrite-sections.ld spill.o -o overwrite-sections --enable-non-contiguous-regions 2>&1 |\
```
omit `-o`

MaskRay commented on lld/ELF/OutputSections.cpp:
```diff
@@ -226,6 +226,13 @@ void OutputSection::finalizeInputSections() {
         isd->sections.push_back(syn);
+        // The merge synthetic section inherits the potential spill locations of
+        // its first contained section.
+        if (script) {
```
Remove the `if`. `script is non-null

mysterymath replied:
Done.

mysterymath committed:
Remove if check around script; should never be hit

MaskRay commented:
Before merging, it's always beneficial to test it locally.

mysterymath commented:
> Before merging, it's always beneficial to test it locally. rebase or merge main

Yep, sorry about that. I was desirous to use the Github UI to squash and rebase the change, but perhaps its best to do such things manually if they can't be separated from the action of committing. I guess I could have rebased the whole stack onto main and tested locally before hitting the big button; that would have caught this too.
